### PR TITLE
feat: [#667] Optimize dependency order

### PR DESCRIPTION
**Pull request by hwbrzzl**
## 📑 Description

Closes https://github.com/goravel/goravel/issues/667

This pull request introduces enhancements to the `foundation/application.go` file by improving the handling of service provider dependencies and bindings, as well as adding comprehensive test cases to validate these changes. The most significant updates include the introduction of virtual bindings for providers without explicit bindings, improved dependency graph construction, and new test scenarios to ensure correctness and robustness.

### Enhancements to service provider dependency handling:

* **Virtual bindings for providers without explicit bindings**: Added logic to create virtual bindings for service providers that lack explicit bindings but have dependencies or "provide-for" relationships. This ensures these providers are included in the dependency graph. (`foundation/application.go`)
* **Improved dependency graph construction**: Enhanced the sorting logic to account for virtual bindings and ensure providers are sorted in a valid topological order. This includes changes to how dependencies and "provide-for" relationships are processed. (`foundation/application.go`) [[1]](diffhunk://#diff-fc0e8c2ae3af78aa10c8fc5dbacc4fb74d177609882d9e2933363cd92358c3b7R399-R459) [[2]](diffhunk://#diff-fc0e8c2ae3af78aa10c8fc5dbacc4fb74d177609882d9e2933363cd92358c3b7L478-R515)

### Debugging and testing improvements:

* **Debugging support**: Added a call to `debug.Dump` to output the configured service providers for easier debugging. (`foundation/application.go`)
* **New test scenarios for edge cases**: Introduced multiple test cases to validate scenarios such as circular dependencies, empty bindings with dependencies or "provide-for" relationships, and unidirectional dependencies. (`foundation/application_test.go`) [[1]](diffhunk://#diff-6557bfe8e38b58d9171f71a99a6402529163c1ebc89309419e4f62e5cf68855eR444) [[2]](diffhunk://#diff-6557bfe8e38b58d9171f71a99a6402529163c1ebc89309419e4f62e5cf68855eR652-R673) [[3]](diffhunk://#diff-6557bfe8e38b58d9171f71a99a6402529163c1ebc89309419e4f62e5cf68855eR993-R1085)
* **Topological order validation**: Added a helper function `isTopologicalOrder` to ensure that the sorted providers maintain a valid topological order. This is used in test cases to verify correctness. (`foundation/application_test.go`)

These changes improve the robustness and maintainability of the service provider sorting mechanism, ensuring proper handling of complex dependency scenarios while providing better debugging and testing capabilities.

## ✅ Checks

- [x] Added test cases for my code

### CODE REVIEW

File: foundation/application.go [hunk 1]
```diff
@@ -475,7 +512,7 @@ func sortConfiguredServiceProviders(providers []foundation.ServiceProvider) []fo
 	// Add any remaining providers that weren't in the dependency graph
 	for _, provider := range providers {
 		if !used[provider] {
-			sortedProviders = append([]foundation.ServiceProvider{provider}, sortedProviders...)
+			sortedProviders = append(sortedProviders, provider)
```
Comment (hwbrzzl) [inline foundation/application.go:512]: The providers that don't set a relationship will be set to the bottom.
